### PR TITLE
`MaterialRateAuxKernel` code, tests, and documentation. 

### LINKED ISSUE #23144: Add auxKernel to compute rate of change in material property over time
## Reason
I want to get the rate of change for a material property.  The TimeDerivative Auxkernel won't work because this material's time derivative isn't computed using AD.  

## Design
I will derive another base class off of MaterialAuxBase that will initialize prop_old and then each rate class derived off this will use prop-prop_old/dt to get the rate.

### PR DESCRIPTION
**Pull request by lynnmunday**
closes #23144

### CODE REVIEW

**Comment by lynnmunday** [comment]
@dschwen will you review?

File: framework/doc/content/source/auxkernels/MaterialRateRealAux.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# MaterialRateRealAux
+
+!syntax description /AuxKernels/MaterialRateRealAux
+
+## Description
+
+The `MaterialRateRealAux` AuxKernel is used to output the rate of change of
+material properties as an element-level, constant variable.
```
Comment (dschwen) [inline framework/doc/content/source/auxkernels/MaterialRateRealAux.md:8]: doesn't have to be constant. could be a higher order monomial

File: framework/include/auxkernels/MaterialRateRealAux.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+template <bool is_ad>
+class MaterialRateRealAuxTempl : public MaterialRateAuxBaseTempl<Real, is_ad>
+{
+public:
+  static InputParameters validParams();
+
+  /**
+   * Class constructor.
+   * @param parameters Input parameters for this object
+   */
```
Comment (dschwen) [inline framework/include/auxkernels/MaterialRateRealAux.h:32]:
```suggestion
```

Comment (lynnmunday) [reply framework/include/auxkernels/MaterialRateRealAux.h]: fixed

File: framework/doc/content/source/auxkernels/MaterialRateRealAux.md (hunk 1 above)
Comment (lynnmunday) [reply framework/doc/content/source/auxkernels/MaterialRateRealAux.md]: fixed

**Comment by lynnmunday** [comment]
@dschwen will you review again and merge?

File: framework/doc/content/source/auxkernels/MaterialRateRealAux.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# MaterialRateRealAux
+
+!syntax description /AuxKernels/MaterialRateRealAux
+
+## Description
+
+The `MaterialRateRealAux` AuxKernel is used to output the rate of change of
+material properties as an element-level, variable.
+The computed value will be the volume-averaged quantity over the element.
```
Comment (dschwen) [inline framework/doc/content/source/auxkernels/MaterialRateRealAux.md:9]: only true for constant monomial

File: framework/include/auxkernels/MaterialRateRealAux.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// MOOSE includes
+#include "MaterialRateAuxBase.h"
+
+// Forward Declarations
+template <bool>
+class MaterialRateRealAuxTempl;
```
Comment (dschwen) [inline framework/include/auxkernels/MaterialRateRealAux.h:17]: what is this forward declaration needed for?

File: framework/include/auxkernels/MaterialRateRealAux.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+typedef MaterialRateRealAuxTempl<false> MaterialRateRealAux;
```
Comment (dschwen) [inline framework/include/auxkernels/MaterialRateRealAux.h:18]: put this at the bottom with the AD typedef

File: framework/src/auxkernels/MaterialRateRealAux.C [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+registerMooseObject("MooseApp", MaterialRateRealAux);
+registerMooseObject("MooseApp", ADMaterialRateRealAux);
+
+template <bool is_ad>
+InputParameters
+MaterialRateRealAuxTempl<is_ad>::validParams()
+{
+  InputParameters params = MaterialAuxBaseTempl<Real, is_ad>::validParams();
+  params.addClassDescription("Outputs element volume-averaged material properties rate of change");
```
Comment (dschwen) [inline framework/src/auxkernels/MaterialRateRealAux.C:20]: where is that "volume averaged" stuff coming from. This is misleading. I'd just remove it.

File: framework/src/auxkernels/MaterialRealAux.C [hunk 7]
```diff
@@ -31,7 +31,7 @@ template <bool is_ad>
 Real
 MaterialRealAuxTempl<is_ad>::getRealValue()
 {
-  // _prop and _qp are members of a dependent template so they need to be qualified with this->
+  // _prop and _qp are members of a dependent name so they need to be qualified with this->
```
Comment (dschwen) [inline framework/src/auxkernels/MaterialRealAux.C:34]: no, just remove

File: test/tests/auxkernels/material_rate_real/material_rate_real.i [hunk 8]
```diff
@@ -0,0 +1,90 @@
+[Mesh]
+  type = GeneratedMesh
+  dim = 2
+  nx = 1
+  ny = 1
+[]
+
+[AuxVariables]
+  [./rate]
```
Comment (dschwen) [inline test/tests/auxkernels/material_rate_real/material_rate_real.i:9]: please use the up to date HIT syntax for new input files

File: test/tests/auxkernels/material_rate_real/tests [hunk 9]
```diff
@@ -0,0 +1,10 @@
+[Tests]
+  design = 'source/auxkernels/MaterialRateRealAux.md'
+  issues = '#23144'
+  [scalar_rate]
+    type = 'CSVDiff'
+    input = 'material_rate_real.i'
+    csvdiff = 'material_rate_real_out.csv'
+    requirement = 'The system shall compute the rate of change of a scalar material property'
```
Comment (dschwen) [inline test/tests/auxkernels/material_rate_real/tests:8]: as a reminder `'` strings are not reflowed, only `"` strings are. This doesn;t matter here, but for long requirement strings it does. I'd recommend to default to using `"` strings, _unless_ you absolutely do not want them to get reflowed.

File: framework/include/auxkernels/MaterialRateAuxBase.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+class MaterialRateAuxBaseTempl : public MaterialAuxBaseTempl<RT, is_ad>
+{
+public:
+  static InputParameters validParams();
+
+  /**
+   * Class constructor
+   * @param parameters The input parameters for this object
+   */
+  MaterialRateAuxBaseTempl(const InputParameters & parameters);
```
[the remark this answers is not in the document]
Comment (lynnmunday) [reply framework/include/auxkernels/MaterialRateAuxBase.h:29]: done

**Comment by lynnmunday** [comment]
@dschwen will you review again.  Can you easily squash the commits when you merge?

**Review by dschwen — APPROVED** [review]

**Pull request merged**